pkg/storagev2/leveldb/store.go

```diff
@@ -0,0 +1,240 @@
+	keys := util.BytesPrefix([]byte(key.Namespace() + forwardSlash))
+	iter := s.DB.NewIterator(keys, nil)
+
+	var c int
+	for iter.Next() {
+		c++
+	}
+
+	iter.Release()
+
+	return c, iter.Error()
+}
+
+func (s *store) Put(item storageV2.Item) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	value, err := item.Marshal()
+	if err != nil {
+		return fmt.Errorf("failed serializing: %w", err)
+	}
+
+	key := []byte(item.Namespace() + forwardSlash + item.ID())
+	return s.DB.Put(key, value, &opt.WriteOptions{Sync: true})
+}
+
+func (s *store) Get(item storageV2.Item) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(item.Namespace() + forwardSlash + item.ID())
+
+	val, err := s.DB.Get(key, nil)
+
+	if errors.Is(err, leveldb.ErrNotFound) {
+		return storageV2.ErrNotFound
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if err = item.Unmarshal(val); err != nil {
+		return fmt.Errorf("failed decoding value %w", err)
+	}
+
+	return nil
+}
+
+func (s *store) Has(sKey storageV2.Key) (bool, error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(sKey.Namespace() + forwardSlash + sKey.ID())
+
+	return s.DB.Has(key, nil)
+}
+
+func (s *store) GetSize(sKey storageV2.Key) (int, error) {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(sKey.Namespace() + forwardSlash + sKey.ID())
+
+	val, err := s.DB.Get(key, nil)
+
+	if errors.Is(err, leveldb.ErrNotFound) {
+		return 0, storageV2.ErrNotFound
+	}
+
+	if err != nil {
+		return 0, err
+	}
+
+	return len(val), nil
+}
+
+func (s *store) Delete(item storageV2.Key) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	key := []byte(item.Namespace() + forwardSlash + item.ID())
+
+	return s.DB.Delete(key, &opt.WriteOptions{Sync: true})
+}
+
+func (s *store) Iterate(q storageV2.Query, fn storageV2.IterateFn) error {
+	s.closeLk.RLock()
+	defer s.closeLk.RUnlock()
+
+	if err := q.Validate(); err != nil {
+		return fmt.Errorf("failed iteration: %w", err)
+	}
+
+	var retErr *multierror.Error
+
+	namespace := q.Factory().Namespace()
+	prefix := namespace + forwardSlash
+
+	keys := util.BytesPrefix([]byte(prefix))
+
+	iter := s.DB.NewIterator(keys, nil)
+
+	nextF := iter.Next
+
+	if q.Order == storageV2.KeyDescendingOrder {
+		nextF = func() bool {
+			nextF = iter.Prev
+			return iter.Last()
+		}
+	}
+
+	for nextF() {
+		nextKey := string(iter.Key())
+		nextVal := iter.Value()
+
+		key := strings.TrimPrefix(nextKey, prefix)
+
+		if filters(q.Filters).matchAny(key, nextVal) {
+			continue
+		}
+
+		var err error
+
+		var res *storageV2.Result
+		switch q.ItemAttribute {
+		case storageV2.QueryItemID, storageV2.QueryItemSize:
+			res = &storageV2.Result{ID: key, Size: len(nextVal)}
+		case storageV2.QueryItem:
+			newItem := q.Factory()
+			err = newItem.Unmarshal(nextVal)
+			res = &storageV2.Result{Entry: newItem}
+		}
+
+		if err != nil {
+			retErr = multierror.Append(retErr, fmt.Errorf("failed unmarshaling: %w", err))
+			break
+		}
+
+		if res == nil {
```

Comment:
Not sure if this will ever happen.

Reply:
this is just a future protection in case we add a new `Query` kind that doesn't match any switch cases.
this way it will return a meaningful error instead of panicking on line `:207`